CASSANDRA-14572 Expose all table metrics in virtual tables [without annotation processor]

Mmuzaf opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/BatchMetricsTable.java:
```diff
@@ -61,6 +67,9 @@ public DataSet data()
         addRow(result, PARTITIONS_PER_UNLOGGED_BATCH, metrics.partitionsPerUnloggedBatch.getSnapshot());
         addRow(result, PARTITIONS_PER_COUNTER_BATCH, metrics.partitionsPerCounterBatch.getSnapshot());
 
+        ClientWarn.instance.warn(String.format("The \"%s\" virtual table is deprecated. " +
```
this is no aligned.

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CQLMetricsTable.java:
```diff
@@ -70,6 +75,8 @@ public DataSet data()
         addRow(result, REGULAR_STATEMENTS_EXECUTED, cqlMetrics.regularStatementsExecuted.getCount());
 
+        ClientWarn.instance.warn("The \"" + TABLE_NAME + "\" virtual table is deprecated. " +
+                                 "Please, use \"system_metrics.clq_group\" virtual table instead.");
```
clq -> cql

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java:
```diff
@@ -0,0 +1,589 @@
+
+        List<AbstractType<?>> partitionKeyTypes = new ArrayList<>(walker.count(Column.Type.PARTITION_KEY));
+        walker.visitMeta(
+            new RowWalker.MetadataVisitor()
```
should not this be again aligned? it should start on the same line as visitMeta

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java:
```diff
@@ -0,0 +1,589 @@
+                                                                                     ColumnFilter columnFilter)
+            {
+                NavigableMap<DecoratedKey, NavigableMap<Clustering<?>, Row>> partitionMap = new ConcurrentSkipListMap<>(
+                    DecoratedKey.comparator);
```
I would violate 120 chars rule if it just reads better, should be on the same line as previous one

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java:
```diff
@@ -0,0 +1,589 @@
+                                   .stream()
+                                   .map(entry -> new DataRowUnfilteredIterator(entry.getKey(),
+                                                                               dataRange.clusteringIndexFilter(
+                                                                                   entry.getKey()),
```
same

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java:
```diff
@@ -0,0 +1,589 @@
+            @Override
+            public T get()
+            {
+                return value == null ? value = delegate.get() : value;
```
btw IDEA is not happy about this, we are assigning and returning, it tells to split this 

    if (value == null)
        value = delegate.get()

    return value

Are we doing this because `value` is `volatile`?

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java:
```diff
@@ -0,0 +1,589 @@
+    }
+
+    private static UnfilteredPartitionIterator createPartitionIterator(
+        TableMetadata metadata,
```
alignement

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java:
```diff
@@ -0,0 +1,589 @@
+
+                        // Push down the column filter to the walker, so we don't have to process the value if it's not queried
+                        ColumnMetadata cm = columnMetas.computeIfAbsent(columnName, name -> metadata.getColumn(
+                            ByteBufferUtil.bytes(name)));
```
just put it on same line

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/SystemViewsKeyspace.java:
```diff
@@ -35,7 +38,13 @@ private SystemViewsKeyspace()
-                    .add(new ThreadPoolsTable(VIRTUAL_VIEWS))
+                    // Fully backward/forward compatible with the legace ThreadPoolsTable under the same "system_views.thread_pools" name.
+                    .add(CollectionVirtualTableAdapter.create(VIRTUAL_VIEWS,
+                            "thread_pools",
```
should be aligned?

Mmuzaf replied:
Yeah, fixed.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/VirtualKeyspace.java:
```diff
@@ -38,6 +41,15 @@ public VirtualKeyspace(String name, Collection<VirtualTable> tables)
         this.tables = ImmutableList.copyOf(tables);
 
+        List<String> duplicates = tables.stream()
+                .map(VirtualTable::name)
```
should be aligned

Mmuzaf replied:
Fixed as it was possible :-)

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java:
```diff
@@ -0,0 +1,589 @@
+                                                                                      rows));
+    }
+
+    /** {@inheritDoc} */
```
why do we have these inheritDoc here? e.g. IDEA just perfectly renders javadoc for this, taking it from the signature of method this overrides. I do not think this is necessary, there is a lot of instances like this in this patch

Mmuzaf replied:
Removed all such comments, I guess this is my old habit kicking in ;-)

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/model/CounterMetricRow.java:
```diff
@@ -0,0 +1,58 @@
+
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+
```
nit: i see there are two lines between import and the comment, why not just one, I think this was covered in my patch

Mmuzaf replied:
Fixed here and for all other places.

smiklosovic commented on src/java/org/apache/cassandra/db/virtual/walker/MetricGroupRowWalker.java:
```diff
@@ -0,0 +1,62 @@
+        switch (type)
+        {
+            case PARTITION_KEY:
+                return 1;
```
you can group PARTITION_KEY with REGULAR because it is both 1, just clustering is 0

Mmuzaf replied:
Fixed.

smiklosovic commented on src/java/org/apache/cassandra/metrics/CassandraMetricsRegistry.java:
```diff
@@ -18,85 +18,324 @@
+            final String groupPrefix = DefaultNameFactory.GROUP_NAME + '.' + groupName + '.';
+            builder.add(createSinglePartitionedKeyFiltered(VIRTUAL_METRICS,
+                                                           METRICS_GROUP_POSTFIX.apply(
+                                                               groupName),
```
I think this would read better if it was just on the same line, same case below

Mmuzaf replied:
Fixed.

smiklosovic commented on test/unit/org/apache/cassandra/cql3/CQLTester.java:
```diff
@@ -476,6 +482,12 @@ public void afterTest() throws Throwable
+    protected static void addMetricsKeyspace()
+    {
+        VirtualKeyspaceRegistry.instance.register(new VirtualKeyspace(VIRTUAL_METRICS,
+                CassandraMetricsRegistry.createMetricsKeyspaceTables()));
```
I guess it should be aligned

Mmuzaf replied:
Fixed.

smiklosovic commented on test/unit/org/apache/cassandra/cql3/CQLTester.java:
```diff
@@ -1763,6 +1775,96 @@ protected void assertRowsNet(ProtocolVersion protocolVersion, ResultSet result,
+                continue;
+
+            Assert.fail(String.format("Rows do not match column definitions. Expected %d columns but got %d",
+                    result.getColumnDefinitions().size(), row.length));
```
alignement

Mmuzaf replied:
Fixed.

Mmuzaf committed:
Expose all dropwizard metrics in virtual tables

patch by Maxim Muzafarov; reviewed by Chris Lohfink, Stefan Miklosovic for CASSANDRA-14572
